hack/generator/pkg/codegen/storage/storage_type_factory.go

```diff
@@ -161,31 +196,56 @@ func (f *StorageTypeFactory) injectConversions(definition astmodel.TypeDefinitio
 		return nil, nil
 	}
 
+	return f.injectConversionsBetween(definition, nextDef)
+}
+
+// inject conversion methods between the two specified definitions
+// upstreamDef is the definition further away from our hub type in our directed conversion graph
+// downstreamDef is the definition closer to our hub type in our directed conversion graph
+func (f *StorageTypeFactory) injectConversionsBetween(
+	upstreamDef astmodel.TypeDefinition, downstreamDef astmodel.TypeDefinition) (*astmodel.TypeDefinition, error) {
+
 	// Create conversion functions
-	knownTypes := make(astmodel.Types)
-	knownTypes.AddTypes(f.inputTypes)
-	knownTypes.AddTypes(f.outputTypes)
+	knownTypes := f.referenceTypes.Copy()
+	knownTypes = knownTypes.OverlayWith(f.outputTypes)
```

Comment:
This is throwing me for a loop a bit. The call chain is:
1. `f.outputTypes.Process(f.injectConversions)` line 106
2. `f.injectConversionsBetween(definition, nextDef)` line 199

and inside `injectConversionsBetween` we then build a set of `knownTypes` (which is `f.referenceTypes + f.outputTypes`) which is then provided to `PropertyConversionContext` for the purposes of looking up types for the conversion.

I understand the objective (build a `knownTypes` collection that we use for lookup), but how it's being done is confusing me a bit:
1. We recompute this set for each invocation of `f.injectConversions`, so we're O(N^2).
2. Since this is reusing the same `f.outputTypes` in various locations and functions you have to pay close attention to temporally where this is called "what is in `outputTypes` at the time this function is invoked"? This took me a second to understand and would be cleaner IMO if the set in question had just been a parameter rather than a struct variable that's constantly being mutated.

I think r.e. (2) above that I had raised a similar comment in a previous PR and you pointed out that some of that was required because of the signature of `Process`. I'm wondering if the usage of `Process` is helping or hurting our complexity here. I can't quite imagine what the code would look like more purely-functional but I'm wondering if the few extra lines of loops would be offset by clearer method signatures and lack of `output` reuse. Bonus that it would also make it easy to just pass in the `knownTypes` from above rather than recomputing it for each element.

Reply:
Technically you can dodge the O(N^2) problem by just precomputing `knownTypes` and stashing it on `f` but I think that leads us further down the "state mutation happening at different times is critical to the code" and also leads to a smell where you have a struct variable that really exists to pass a parameter to a method (and is useless before and after that call)

Reply:
This is why I collapsed `referenceType` and `outputTypes` into the same set - there's no need to recompute it.

Reply:
I think you've resolved the issue with recomputing (item 1 above).

I do still wonder if the `types.Process(...)` approach which leads itself to a mutable collection (`types`) stored on the `StorageTypeFactory` is hurting us more than it's helping though, and if a set of pure (or more-pure) functions would make the sequence of mutations clearer even if we had to do away with `types.Process(...)` and just write a loop.

Reply:
I've gotten rid of `storage_types_factory`, replacing it with a set of individual pipeline stages.